File: 904.fruit-into-baskets.java
/*
 * @lc app=leetcode id=904 lang=java
 *
 * [904] Fruit Into Baskets
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

class Solution {
    public int totalFruit(int[] fruits) {
        Map<Integer, Integer> map = new HashMap<>();
        int ret = 0;
        for (int left = 0, right = 0; right != fruits.length; ++right) {
            map.put(fruits[right], map.getOrDefault(fruits[right], 0) + 1);
            while (map.size() > 2) {
                map.put(fruits[left], map.get(fruits[left]) - 1);
                map.remove(fruits[left], 0);
                left += 1;
            }
            ret = Math.max(ret, right - left + 1);
        }
        return ret;
    }
}
// @lc code=end

File: 904.fruit-into-baskets.cpp
/*
 * @lc app=leetcode id=904 lang=cpp
 *
 * [904] Fruit Into Baskets
 */

// @lc code=start

#include <vector>
#include <algorithm>
#include <map>

using namespace std;

class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int l = 0, r = 1;
        map<int, int> basket;   // type -> count
        int maxCnt = 0;
        basket[fruits.at(l)] = 1;
        while (r <= fruits.size()) {
            while (basket.size() > 2) {
                basket[fruits.at(l)] -= 1;
                if (basket[fruits.at(l)] == 0) {
                    basket.erase(fruits.at(l));
                }
                l += 1;
            }
            maxCnt = max(maxCnt, r - l);
            if (r == fruits.size()) {
                break;
            }
            basket[fruits.at(r)] += 1;
            r += 1;
        }
        return maxCnt;
    }
};
// @lc code=end
